### ISSUE TITLE
function upload order breaks when dependencies exist

### ISSUE THREAD

**Issue by eknowles** [Plazide/fauna-gql-upload#5]
I've noticed a bug when trying to upload a function that calls another function.
When I clean my database and re run the upload I do get various errors like `validation failed: document data is not valid.
` these errors end up referencing nearly all the files.

I traced this issue back to a new function I had written that called another function. Like so.

Function A
```
Call(Function('function_b'))
```

From what I can tell, since the functions are being uploaded in any order, if function A is uploaded before function B, it means validation will fail, and will also fail for all other roles or functions that might reference function a.

A solution to this would be.

a. Static analysis on the function body and detect this dependency then update the code to upload files sorted by this dependency.
b. The developer can add a `priority: number` in the export of the function, this is easy but also means the code would need to read all files at once.
c. Rename files so they are in the right order alphabetically. this might be adding a number prefix.

number prefix
```
./functions/1-function-b
./functions/2-function-a
./functions/3-some-other-function
```

**Comment by Plazide** [comment]
That's interesting. I thought the root cause of this was the role files. When you upload a function that references a role that doesn't exist, it will fail. And when you upload a role referencing a function that doesn't exist, it will fail.

Referencing non-existent resources in the function body doesn't seem to be an issue, I can't produce an error.

I created a fresh database and setup a new project with your fork of `fauna-gql-upload`. I created a function called `test` that references a function that has not been created, like this:
```js
const { query: q } = require("faunadb");

module.exports = {
	name: "test",
	body: 
	q.Query(
		q.Lambda(
			"_",
			q.Call(q.Function("i-do-not-exist"))
		)
	)
}
```

When calling the upload command, I get the following output:
![bild](https://user-images.githubusercontent.com/20641118/88379776-c82d5980-cda3-11ea-9e3e-afa3012565c9.png)


If I then add a role to that function:
```js
const { query: q } = require("faunadb");

module.exports = {
	name: "test",
	role: q.Role("i-do-not-exist"),
	body: 
	q.Query(
		q.Lambda(
			"_",
			q.Call(q.Function("i-do-not-exist"))
		)
	)
}
```

I get something like what you describe:
![bild](https://user-images.githubusercontent.com/20641118/88380999-d8463880-cda5-11ea-98ee-f6befbbc0612.png)


In other words, I don't get errors when referencing functions in the function body, but I do get errors when referencing a role on the role field. So, I believe the issue is caused by roles not existing and not by referencing non-existent functions. This would explain the error you are getting since the roles are uploaded last of all the resources.

A solution to this might be to first create functions without roles and then, after uploading the roles, populate the role field.

**Comment by Plazide** [comment]
@eknowles 

Was I correct, or did I misunderstand the problem?
